PR: https://github.com/odoo/odoo/pull/208005

From: 2fca445d07a803953481e8b4967806e3453a00da
From: Maxime de Neuville
Date: 2025-09-09 12:52:48

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'res.config.settings'

Total Changes: 42

[IMP] project,sale_project,sale_timesheet,web: clean settings

This commit aims to remove some global settings to only keep them in the project form view:

1. More specifically, it removes the global settings "Recurring Tasks", "Task Dependencies" and "Milestones" from the global settings.
And moves "Recurring Tasks" and "Task Dependencies" in the project form view directly. It keeps the "Milestones" settings in the project form view as it was.

* The way groups "project.group_project_milestone", "project.group_project_recurring_tasks" and "project.group_project_recurring_tasks" are given to the user is now different, as the global settings are removed.
It now works like this:
=> When the user enables one of the settings we mentioned (e.g. "Milestones") in any project and the related group (e.g. "group_project_milestone") is not already added to the base user group, it is added.
=> If the user disables the setting from a project/remove a project which had the setting enabled, if after that disabling there is no project (even archived ones) having the setting enabled, then the related group is removed from the base user group.

* The default value given to the setting of a new project is always False.

* We remove the behaviour introduced by this commit: https://github.com/odoo/odoo/commit/50ffc339991b136b48525b2d60aa34a3039bb8da

2. The "Customer Ratings" setting was removed from the global settings and the project form view. It is moved to the task stage form view.

* If the rating status of a stage is set to "stage": when a task reaches that stage, an email will be automatically sent to get the customer rating.
  If the rating status of a stage is set to "periodic": an email will be sent periodically at the given interval as long as the task remains in the stage.
* If sale_project is installed, we only show the setting if at least one of the project of the stage is billable
* The "Rating Email Template" is now required if the setting is enabled.
* Some renaming was made in the setting description to make it clearer.
* The group "project.group_project_rating" is removed so we always make visible the customer rating in all views and in the reporting.

3. When the user enables the top bar a project which only has one visible action (e.g. "Tasks"), it now directly opens the dropdown so that the user will gain time to select the actions to show.

task-4700865

closes odoo/odoo#208005

Related: odoo/enterprise#84657
Related: odoo/upgrade#7640
Signed-off-by: Vincent Larcin (vila) <vila@odoo.com>

================================= pseudo patch: =================================

--- a/addons/sale_project/models/__init__.py
+++ b/addons/sale_project/models/__init__.py
@@ -5,9 +5,9 @@ from . import product_template
 from . import project_milestone
 from . import project_project
 from . import project_task_recurrence
+from . import project_task_type
 from . import project_task
 from . import project_update
-from . import res_config_settings
 from . import sale_order_line
 from . import sale_order_template_line
 from . import sale_order

--- a/None
+++ b/addons/sale_project/models/project_task_type.py
@@ -0,0 +1,17 @@
+from odoo import api, fields, models
+
+
+class ProjectTaskType(models.Model):
+    _inherit = 'project.task.type'
+
+    show_rating_active = fields.Boolean(compute='_compute_show_rating_active', export_string_translation=False)
+
+    @api.depends('project_ids.allow_billable')
+    def _compute_show_rating_active(self):
+        for stage in self:
+            stage.show_rating_active = any(stage.project_ids.mapped('allow_billable'))
+
+    @api.onchange('project_ids')
+    def _onchange_project_ids(self):
+        if not any(self.project_ids.mapped('allow_billable')):
+            self.rating_active = False

--- a/addons/sale_project/models/res_config_settings.py
+++ b/None
@@ -1,22 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import models
-
-
-class ResConfigSettings(models.TransientModel):
-    _inherit = 'res.config.settings'
-
-    def set_values(self):
-        super().set_values()
-        if self.group_project_milestone:
-            # Search the milestones containing a SOL and change the qty_delivered_method field of the SOL and the
-            # service_policy field set on the product to convert from manual to milestones.
-            milestones = self.env['project.milestone'].search_fetch([('sale_line_id', '!=', False)], ['sale_line_id'])
-            sale_lines = milestones.sale_line_id.sudo()
-            sale_lines.product_id.service_policy = 'delivered_milestones'
-        else:
-            product_domain = [('type', '=', 'service'), ('service_type', '=', 'milestones')]
-            products = self.env['product.product'].search(product_domain)
-            products.service_policy = 'delivered_manual'
-            self.env['sale.order.line'].sudo().search([('product_id', 'in', products.ids)]).qty_delivered_method = 'manual'

--- a/addons/sale_project/models/sale_order_line.py
+++ b/addons/sale_project/models/sale_order_line.py
@@ -182,6 +182,7 @@ class SaleOrderLine(models.Model):
             'company_id': self.company_id.id,
             'allow_billable': True,
             'user_id': self.product_id.project_template_id.user_id.id,
+            'allow_milestones': self.product_id.service_type == 'milestones',
         }
 
     def _timesheet_create_project(self):
